File: PHP/LatLon.php
<?php
/* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -  *
 *  Vector-based spherical geodetic (latitude/longitude) functions    (c) Chris Veness 2011-2016  *
 *                                                                                   MIT Licence  *
 * www.movable-type.co.uk/scripts/latlong-vectors.html                                            *
 * www.movable-type.co.uk/scripts/geodesy/docs/module-latlon-nvector-spherical.html               *
 * - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -  */

require 'Vector3d.php';


/**
 * Tools for working with points and paths on (a spherical model of) the earth’s surface using a
 * vector-based approach using 'n-vectors' (rather than the more common spherical trigonometry:
 * a vector-based approach makes many calculations much simpler, and easier to follow, compared
 * with trigonometric equivalents).
 *
 * Note on a spherical model earth, an n-vector is equivalent to a normalized version of an (ECEF)
 * cartesian coordinate.
 *
 * @module   latlon-vectors
 * @requires Vector3d
 */
class LatLon
{
    //Instance Variables
    private $lat = 0.0;
    private $lon = 0.0;

    /**
     * Creates a LatLon point on spherical model earth.
     *
     * @constructor
     * @param {number} lat - Latitude in degrees.
     * @param {number} lon - Longitude in degrees.
     *
     * @example
     *   var p1 = new LatLon(52.205, 0.119);
     */
    function LatLon($lat, $lon)
    {
        $this->lat = $lat;
        $this->lon = $lon;
    }


    /**
     * Converts ‘this’ lat/lon point to Vector3d n-vector (normal to earth's surface).
     *
     * @returns {Vector3d} Normalized n-vector representing lat/lon point.
     *
     * @example
     *   var p = new LatLon(45, 45);
     *   var v = p->toVector(); // [0.5000,0.5000,0.7071]
     */
    function toVector()
    {
        $rLat = deg2rad($this->lat);
        $rLon = deg2rad($this->lon);

        $x = cos($rLat) * cos($rLon);
        $y = cos($rLat) * sin($rLon);
        $z = sin($rLat);

        return new Vector3d($x, $y, $z);
    }


    /**
     * N-vector normal to great circle obtained by heading on given bearing from ‘this’ point.
     *
     * Direction of vector is such that initial bearing vector b = c × p.
     *
     * @param   {number}   bearing - Compass bearing in degrees.
     * @returns {Vector3d} Normalized vector representing great circle.
     *
     * @example
     *   var p1 = new LatLon(53.3206, -1.7297);
     *   var gc = p1->greatCircle(96.0); // [-0.794,0.129,0.594]
     */
    function greatCircle($bearing)
    {
        $rLat = deg2rad($this->lat);
        $rLon = deg2rad($this->lon);
        $theta = deg2rad($bearing);

        $x =  sin($rLon) * cos($theta) - sin($rLat) * cos($rLon) * sin($theta);
        $y = -cos($rLon) * cos($theta) - sin($rLat) * sin($rLon) * sin($theta);
        $z =  cos($rLat) * sin($theta);

        return new Vector3d($x, $y, $z);
    }


    /**
     * Returns the distance from ‘this’ point to the specified point.
     *
     * @param   {LatLon} point - Latitude/longitude of destination point.
     * @param   {number} [radius=6371e3] - (Mean) radius of earth (defaults to radius in meters).
     * @returns {number} Distance between this point and destination point, in same units as radius.
     *
     * @example
     *   var p1 = new LatLon(52.205, 0.119);
     *   var p2 = new LatLon(48.857, 2.351);
     *   var d = p1->distanceTo(p2); // 404.3 km
     */
    function distanceTo($point, $radius=NULL)
    {
        if ($radius === NULL)  $radius = 6371000;

        $p1 = $this->toVector();
        $p2 = $point->toVector();

        $delta = $p1->angleTo($p2);
        $d = $delta * $radius;

        return d;
    }


    /**
     * Returns the (initial) bearing from ‘this’ point to the specified point, in compass degrees.
     *
     * @param   {LatLon} point - Latitude/longitude of destination point.
     * @returns {number} Initial bearing in degrees from North (0°..360°).
     *
     * @example
     *   var p1 = new LatLon(52.205, 0.119);
     *   var p2 = new LatLon(48.857, 2.351);
     *   var b1 = p1->bearingTo(p2); // 156.2°
     */
    function bearingTo($point)
    {
        $p1 = $this->toVector();
        $p2 = $point->toVector();

        $northPole = new Vector3d(0.0, 0.0, 1.0);

        $c1 = $p1->cross($p2);        // great circle through p1 & p2
        $c2 = $p1->cross($northPole); // great circle through p1 & north pole

        // bearing is (signed) angle between c1 & c2
        $bearing = rad2deg( $c1->angleTo($c2, $p1) );

        return ($bearing + 360) % 360; // normalize to [0, 360]
    }


    /**
     * Returns the midpoint between ‘this’ point and specified point.
     *
     * @param   {LatLon} point - Latitude/longitude of destination point.
     * @returns {LatLon} Midpoint between this point and destination point.
     *
     * @example
     *   var p1 = new LatLon(52.205, 0.119);
     *   var p2 = new LatLon(48.857, 2.351);
     *   var pMid = p1->midpointTo(p2); // 50.5363°N, 001.2746°E
     */
    function midpointTo($point)
    {
        $p1 = $this->toVector();
        $p2 = $point->toVector();

        $mid = $p1->plus($p2)->unit();

        return $mid->toLatLonS();
    }


    /**
     * Returns the destination point from ‘this’ point having travelled the given distance on the
     * given initial bearing (bearing will normally vary before destination is reached).
     *
     * @param   {number} distance - Distance travelled, in same units as earth radius (default: meters).
     * @param   {number} bearing - Initial bearing in degrees from north.
     * @param   {number} [radius=6371e3] - (Mean) radius of earth (defaults to radius in meters).
     * @returns {LatLon} Destination point.
     *
     * @example
     *   var p1 = new LatLon(51.4778, -0.0015);
     *   var p2 = p1->destinationPoint(7794, 300.7); // 51.5135°N, 000.0983°W
     */
    function destinationPoint($distance, $bearing, $radius=NULL)
    {
        if ($radius === NULL)  $radius = 6371000;

        $delta = $distance / $radius; // angular distance in radians

        // get great circle obtained by starting from 'this' point on given bearing
        $c = $this->greatCircle($bearing);

        $p1 = $this->toVector();

        $x = $p1->times(cos($delta));           // component of p2 parallel to p1
        $y = $c->cross(p1)->times(sin($delta)); // component of p2 perpendicular to p1

        $p2 = $x->plus($y)->unit();

        return $p2->toLatLonS();
    }


    /**
     * Returns (signed) distance from ‘this’ point to great circle defined by start-point and end-point/bearing.
     *
     * @param   {LatLon}        pathStart - Start point of great circle path.
     * @param   {LatLon|number} pathBrngEnd - End point of great circle path or initial bearing from great circle start point.
     * @param   {number}        [radius=6371e3] - (Mean) radius of earth (defaults to radius in meters).
     * @returns {number}        Distance to great circle (-ve if to left, +ve if to right of path).
     *
     * @example
     *   var pCurrent = new LatLon(53.2611, -0.7972);
     *
     *   var p1 = new LatLon(53.3206, -1.7297), brng = 96.0;
     *   var d = pCurrent->crossTrackDistanceTo(p1, brng);// -305.7 m
     *
     *   var p1 = new LatLon(53.3206, -1.7297), p2 = new LatLon(53.1887, 0.1334);
     *   var d = pCurrent->crossTrackDistanceTo(p1, p2);  // -307.5 m
     */
    function crossTrackDistanceTo($pathStart, $pathBrngEnd, $radius=NULL)
    {
        if ($radius === NULL)  $radius = 6371000;

        $p = $this->toVector();

        if ( isinstance(pathBrngEnd, LatLon) )
            // great circle defined by two points
            $gc = $pathStart->toVector()->cross($pathBrngEnd->toVector());
        else
            // great circle defined by point + bearing
            $gc = $pathStart->greatCircle($pathBrngEnd);

        // (signed) angle between point & great circle normal vector
        $alpha = $gc->angleTo($p, $p->cross($gc)); 
        $alpha = ($alpha < 0 ? -M_PI : M_PI) / 2 - $alpha;

        $d = $alpha * $radius;

        return $d;
    }


    /**
     * String representation of LatLon point.
     *
     * @param   {number} [precision=5] - Number of decimal places to be used.
     * @returns {string} LatLon represented as (lat, lon).
     */
    function __toString()
    {
        return "({$this->lat}, {$this->lon})";
    }
}


/**
 * Converts ‘this’ (geocentric) cartesian vector to (spherical) latitude/longitude point.
 *
 * @returns  {LatLon} Latitude/longitude point vector points to.
 *
 * @example
 *   var v = new Vector3d(0.500, 0.500, 0.707);
 *   var p = v->toLatLonS(); // 45.0°N, 45.0°E
 */
function toLatLonS()
{
    $rLat = atan2($this->z, sqrt($this->x*$this->x + $this->y*$this->y));
    $rLon = atan2($this->y, $this->x);
    return new LatLon(rad2deg($rLat), rad2deg($rLon));
}

// Assign toLatLonS function prototype to Vector3d class
//Vector3d->toLatLonS = toLatLonS;
?>

File: latlon.py
# -*- coding: utf-8 -*-

"""
Vector-based spherical geodetic (latitude/longitude) functions
(c) Chris Veness 2011-2016 MIT License

www.movable-type.co.uk/scripts/latlong-vectors.html
www.movable-type.co.uk/scripts/geodesy/docs/module-latlon-nvector-spherical.html
"""

import math
from typing import Union

from vector3d import Vector3d


class LatLon(object):
    """
    Tools for working with points and paths on (a spherical model of) the
    earth’s surface using a vector-based approach using 'n-vectors' (rather
    than the more common spherical trigonometry: a vector-based approach makes
    many calculations much simpler, and easier to follow, compared with
    trigonometric equivalents).

    Note on a spherical model earth, an n-vector is equivalent to a normalized
    version of an (ECEF) * cartesian coordinate.
    """

    def __init__(self, lat: float, lon: float):
        """
        Creates a LatLon point on spherical model earth.
        :param lat: float Latitude in degrees.
        :param lon: float Longitude in degrees.

        >>> p1 = LatLon(52.205, 0.119)
        """
        self.lat = float(lat)
        self.lon = float(lon)

    def __str__(self) -> str:
        """
        String representation of LatLon point.
        :rtype: str
        :return: LatLon represented as (lat, lon).
        """
        return "({self.lat:5f}, {self.lon:5f})".format(self=self)

    def __repr__(self):
        return "<LatLon(lat={self.lat:5f}, lon={self.lon:5f})".format(self=self)

    def to_vector(self) -> Vector3d:
        """
        Converts 'self' to Vector3d n-vector (normal to earth's surface).
        :rtype: Vector3d
        :return: Normalized n-vector representing lat/lon point.

        >>> p = LatLon(45, 45)
        >>> p.to_vector()
        <Vector3d(x=0.500000, 0.500000, 0.707107)>
        """
        r_lat = math.radians(self.lat)
        r_lon = math.radians(self.lon)

        x = math.cos(r_lat) * math.cos(r_lon)
        y = math.cos(r_lat) * math.sin(r_lon)
        z = math.sin(r_lat)

        return Vector3d(x, y, z)

    def great_circle(self, bearing: float) -> Vector3d:
        """
        N-vector normal to great circle obtained by heading on given bearing
        from 'self'.
        :param bearing: Compass bearing in degrees.
        :rtype: Vector3d
        :return: Normalized vector representing great circle.

        >>> p1 = LatLon(53.3206, -1.7297)
        >>> p1.great_circle(96.0)
        <Vector3d(x=-0.794079, 0.128556, 0.594065)>
        """
        r_lat = math.radians(self.lat)
        r_lon = math.radians(self.lon)
        theta = math.radians(bearing)

        x = math.sin(r_lon) * math.cos(theta) \
            - math.sin(r_lat) * math.cos(r_lon) * math.sin(theta)
        y = -math.cos(r_lon) * math.cos(theta) \
            - math.sin(r_lat) * math.sin(r_lon) * math.sin(theta)
        z = math.cos(r_lat) * math.sin(theta)

        return Vector3d(x, y, z)

    def distance_to(self, point: 'LatLon', radius: int = 6371e3) -> float:
        """
        Returns the distance from 'self' to the specified point.
        :param point: Latitude/longitude of destination point.
        :param radius: (Mean) radius of earth (defaults to radius in meters).
        :rtype: float
        :return: Distance between this point and destination point, in same
        units as radius.

        >>> p1 = LatLon(52.205, 0.119)
        >>> p2 = LatLon(48.857, 2.351)
        >>> p1.distance_to(p2)
        404279.16398867936
        """
        p1 = self.to_vector()
        p2 = point.to_vector()

        delta = p1.angleTo(p2)
        d = delta * radius

        return d

    def bearing_to(self, point: 'LatLon') -> float:
        """
        Returns the (initial) bearing from 'self'to the specified point, in
        compass degrees.
        :param point: Latitude/longitude of destination point.
        :rtype: float
        :return: Initial bearing in degrees from North (0°..360°).

        >>> p1 = LatLon(52.205, 0.119)
        >>> p2 = LatLon(48.857, 2.351)
        >>> p1.bearing_to(p2)
        156.16658258153177
        """
        p1 = self.to_vector()
        p2 = point.to_vector()

        north_pole = Vector3d(0.0, 0.0, 1.0)

        c1 = p1.cross(p2)  # great circle through p1 & p2
        c2 = p1.cross(north_pole)  # great circle through p1 & north pole

        # bearing is (signed) angle between c1 & c2
        bearing = math.degrees(c1.angleTo(c2, p1))

        return (bearing + 360) % 360  # normalize to [0, 360]

    def midpoint_to(self, point: 'LatLon') -> 'LatLon':
        """
        Returns the midpoint between 'self' and specified point.
        :param point: Latitude/longitude of destination point.
        :rtype: LatLon
        :return: Midpoint between 'self' point and destination point.

        >>> p1 = LatLon(52.205, 0.119)
        >>> p2 = LatLon(48.857, 2.351)
        >>> p1.midpoint_to(p2)
        <LatLon(lat=50.536327, lon=1.274614)
        """
        p1 = self.to_vector()
        p2 = point.to_vector()

        mid = p1.plus(p2).unit()

        return mid.toLatLonS()

    def destination_point(
        self,
        distance: float,
        bearing: float,
        radius: int = 6371e3
    ) -> 'LatLon':
        """
        Returns the destination point from 'self' having traveled the given
        distance on the given initial bearing (bearing will normally vary
        before destination is reached).
        :param distance: Distance traveled, in same units as earth radius
        (default: meters).
        :param bearing: Initial bearing in degrees from north.
        :param radius: (Mean) radius of earth (defaults to radius in meters).
        :rtype: LatLon
        :return: Destination point.

        >>> p1 = LatLon(51.4778, -0.0015)
        >>> p1.destination_point(7794, 300.7)
        <LatLon(lat=51.513546, lon=-0.098345)
        """
        delta = distance / radius  # angular distance in radians

        # get great circle obtained by starting from 'self' on given bearing
        c = self.great_circle(bearing)

        p1 = self.to_vector()

        # component of p2 parallel to p1
        x = p1.times(math.cos(delta))

        # component of p2 perpendicular to p1
        y = c.cross(p1).times(math.sin(delta))

        p2 = x.plus(y).unit()

        return p2.toLatLonS()

    def cross_track_distance_to(
        self,
        path_start: 'LatLon',
        path_brng_end: Union['LatLon', float],
        radius: int = 6371e3
    ) -> float:
        """
        Returns (signed) distance from 'self' to great circle defined by
        start-point and end-point/bearing.
        :param path_start: Start point of great circle path.
        :param path_brng_end: End point of great circle path or initial bearing
        from great circle start point.
        :param radius: (Mean) radius of earth (defaults to radius in meters).
        :rtype: float
        :return: Distance to great circle (-ve if to left, +ve if to right of
        path).

        >>> pCurrent = LatLon(53.2611, -0.7972)
        >>> p1 = LatLon(53.3206, -1.7297)
        >>> brng = 96.0
        >>> pCurrent.cross_track_distance_to(p1, brng)
        -305.66486077528987

        >>> p1 = LatLon(53.3206, -1.7297)
        >>> p2 = LatLon(53.1887, 0.1334)
        >>> pCurrent.cross_track_distance_to(p1, p2)
        -307.54957041913156
        """
        p = self.to_vector()

        if isinstance(path_brng_end, LatLon):
            # great circle defined by two points
            gc = path_start.to_vector().cross(path_brng_end.to_vector())
        else:
            # great circle defined by point + bearing
            gc = path_start.great_circle(path_brng_end)

        # (signed) angle between point & great circle normal vector
        alpha = gc.angleTo(p, p.cross(gc))

        # (signed) angle between point & great circle
        alpha = (-1 if alpha < 0 else 1) * math.pi / 2 - alpha

        d = alpha * radius

        return d


def to_latlon(self) -> LatLon:
    """
    Converts 'self' (geocentric) cartesian vector to (spherical)
    latitude/longitude point.
    :rtype: LatLon
    :return: Latitude/longitude point vector points to.

    >>> v = Vector3d(0.500, 0.500, 0.707)
    >>> v.to_latlon()
    <LatLon(lat=44.995674, lon=45.000000)
    """
    r_lat = math.atan2(self.z, math.sqrt(self.x * self.x + self.y * self.y))
    r_lon = math.atan2(self.y, self.x)

    return LatLon(math.degrees(r_lat), math.degrees(r_lon))


# Assign to_latlon function prototype to Vector3d class
Vector3d.toLatLonS = to_latlon
